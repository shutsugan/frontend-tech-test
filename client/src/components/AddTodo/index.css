.add-todo {}

.required {
    text-transform: capitalize;
    color: var(--label);
}

.required::after {
    content: '*';
    color: var(--secondary);
}

.field__input {
    border: solid .5px var(--border);
    border-radius: var(--radius);
    outline: none;
    transition: border-color var(--duration) var(--transition);
}

.field__input:focus {
    border-color: var(--main);
}