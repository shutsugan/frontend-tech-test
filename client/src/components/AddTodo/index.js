import React, { useState, useEffect } from 'react';
import { useStoreState, useStoreActions } from 'easy-peasy';

import './index.scss';

const AddTodo = _ => {
    const [id, setId] = useState('');
    const [title, setTitle] = useState('');
    const [description, setDescription] = useState('');

    const selectedTodo = useStoreState(state => state.todos.selectedTodo);

    useEffect(_ => {
        let id, title, description = '';

        if (selectedTodo) {
            id = selectedTodo.id;
            title = selectedTodo.title;
            description = selectedTodo.description;
        }

        setId(id);
        setTitle(title);
        setDescription(description);
    }, [selectedTodo]);

    const addTodo = useStoreActions(actions => actions.todos.addTodo);
    const updateTodo = useStoreActions(actions => actions.todos.updateTodo);
    
    const handleChange = ({ target }, setter) => setter(target.value);

    const handleAddTodo = async _ => {
        !selectedTodo
            ? await addTodo({ title, description })
            : await updateTodo({ id, title, description });

        setId('');
        setTitle('');
        setDescription('');
    };

    return (
        <div className="add-todo full">
            <div className="fleid flex-column start mrb-16 relative">
                <label className="required">Title</label>
                <input
                    aria-label="title"
                    className="field__input full pd-16"
                    value={title}
                    onChange={event => handleChange(event, setTitle)}
                />
            </div>
            <div className="fleid flex-column start mrb-16 relative">
                <label className="required">Description</label>
                <textarea
                    aria-label="description"
                    className="field__input full pd-16 ht-80"
                    value={description}
                    onChange={event => handleChange(event, setDescription)}>                
                </textarea>
            </div>
            <button
                aria-label="button"
                className="button full pd-16"
                onClick={handleAddTodo}>
                Add Todo
            </button>
        </div>
    );
}

export default AddTodo;