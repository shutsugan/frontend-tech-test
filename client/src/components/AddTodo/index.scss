.required {
    text-transform: capitalize;
    color: var(--label);
    &::after {
        content: '*';
        color: var(--secondary);
    }
}

.field__input {
    border: solid 0.5px var(--border);
    border-radius: var(--radius);
    outline: none;
    transition: border-color var(--duration) var(--transition);
    &:focus {
        border-color: var(--main);
    }
}