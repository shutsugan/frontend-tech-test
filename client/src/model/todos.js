import { action, computed, thunk } from "easy-peasy";

const base_url = 'http://localhost:9001';

const noop = _ => {};
export const save = async (axios, actions, method, url, callback = noop) => {
    const res = await axios[method](`${base_url}/task/${url}`);
    const { task, message } = await res.data;

    callback(task);
    actions.setMessage(message);
};

export const checkProps = (actions, title, description) => {
    if (!title && !description) return actions.setMessage('Fields are reuqired');
};

export default {
    todos: [],
    selectedTodo: null,
    message: null,

    list: computed(state => Object.values(state.todos)),

    //actions
    setMessage: action((state, message) => state.message = message),
    selectTodo: action((state, todo) => state.selectedTodo = todo),
    added: action((state, todo) => state.todos[todo.id] = todo),

    fetched: action((state, payload = []) => {
        state.todos = payload.reduce((acc, todo) => {
            acc[todo.id] = todo;

            return acc;
        }, {});
    }),

    //thunks
    fetchTodos: thunk(async (actions, _, { injections }) => {
        const { axios } = injections
        const res = await axios.get(`${base_url}/tasks`);
        const { tasks } = await res.data;

        actions.fetched(tasks);
    }),

    addTodo: thunk(async (actions, { title, description }, { injections }) => {
        const { axios } = injections

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'post',
            `create/${title}/${description}`,
            todo => actions.added(todo)
        );
    }),

    updateTodo: thunk(async (actions, { id, title, description }, { injections }) => {
        const { axios } = injections;

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'put',
            `update/${id}/${title}/${description}`,
            _ => {
                actions.fetchTodos();
                actions.selectTodo(null);
            }
        );
    }),

    removeTodo: thunk(async (actions, id, { injections }) => {
        const { axios } = injections;

        save(
            axios,
            actions,
            'delete',
            `delete/${id}`,
            _ => actions.fetchTodos()
        );
    })
};