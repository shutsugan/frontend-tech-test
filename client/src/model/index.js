import todos from './todos';
import { thunk } from 'easy-peasy';

export default {
    todos,
    initialiseTodos: thunk(async (actions, payload, { dispatch }) => {
        await dispatch.todos.fetchTodos();
    })
};