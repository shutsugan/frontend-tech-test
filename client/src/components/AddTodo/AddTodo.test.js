import React from 'react';
import renderer from 'react-test-renderer';
import { createStore, StoreProvider } from 'easy-peasy';
import { expect as chaiExpect } from 'chai';

import model from '../../model';
import AddTodo from './index';

const store = createStore(model);

describe('AddTodo component', () => {
    it ('should render add todo componenet', () => {
        const tree = renderer.create(
            <StoreProvider store={store}>
                <AddTodo />
            </StoreProvider>
        ).toJSON();
        
        expect(tree).toMatchSnapshot();
        chaiExpect(tree).to.be.an('object');
        chaiExpect(tree).to.have.own.property('props');
        chaiExpect(tree.children.length).to.equal(3);
    });
});