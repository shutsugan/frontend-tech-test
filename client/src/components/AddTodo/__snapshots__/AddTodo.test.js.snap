// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo component should render add todo componenet 1`] = `
<div
  className="add-todo full"
>
  <div
    className="fleid flex-column start mrb-16 relative"
  >
    <label
      className="required"
    >
      Title
    </label>
    <input
      aria-label="title"
      className="field__input full pd-16"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="fleid flex-column start mrb-16 relative"
  >
    <label
      className="required"
    >
      Description
    </label>
    <textarea
      aria-label="description"
      className="field__input full pd-16 ht-80"
      onChange={[Function]}
      value=""
    />
  </div>
  <button
    aria-label="button"
    className="button full pd-16"
    onClick={[Function]}
  >
    Add Todo
  </button>
</div>
`;
